doc/source/object-store.rst

```diff
@@ -0,0 +1,169 @@
+    object_id = ray.put(y)
+
+2. The return values of a remote function.
+
+.. code:: python
+
+    @ray.remote
+    def remote_function():
+        return 1
+
+    # This places `1` into the object store.
+    object_id = remote_function.remote()
+
+
+3. Arguments to remote functions (except for simple arguments like ints or
+   floats).
+
+.. code:: python
+
+    @ray.remote
+    def remote_function(y):
+        # Note that inside the remote function, the actual argument is provided.
+        return len(y)
+
+    argument = [1, 2, 3, 4]
+    # This implicitly places `argument` into the object store.
+    remote_function.remote(argument)
+
+In the single node setting, all Ray workers can read the same object in the object store without copying (zero-copy reads). Once an object is placed in the object store, it is immutable. Any writes to the read-only object will result in a copy into the local process memory.
+
+Distributed Setting
+~~~~~~~~~~~~~~~~~~~
+
+Each node has its own object store. When data is put into a node object store, it
+does not get automatically broadcasted; it remains local until requested by another
+task on another node.
+
+
+Numpy Arrays
+------------
+
+Ray optimizes for numpy arrays by using the `Apache Arrow`_ data format.
+Each numpy array object holds a pointer to the relevant array held in shared memory.
+
+There are some advantages to this form of serialization:
+
+- Deserialization can be very fast.
+- Memory is shared between processes so worker processes can all read the same
+  data without having to copy it.
+
+**Nested Numpy Arrays:** When we deserialize a list of numpy arrays from the object store, we still create a Python list of numpy array objects. However, rather than copy each numpy array,
+
+.. _`Apache Arrow`: https://arrow.apache.org/
+
+Comparison to Pickle
+~~~~~~~~~~~~~~~~~~~~
+
+Pickle is standard Python serialization library. However, for numerical workloads, pickling and unpickling can be inefficient.
+
+For example, if multiple processes want to access a Python list of numpy arrays, each process must unpickle the list and create its own new copies of the arrays. This can lead to high memory overheads, even when all processes are read-only and could easily share memory.
+
+
+Supported Datatypes
+-------------------
+
+Ray does not currently support serialization of arbitrary Python objects.  The
+set of Python objects that Ray can serialize using Arrow includes the following.
+
+1. Primitive types: ints, floats, longs, bools, strings, unicode, and numpy
+   arrays.
+2. Any list, dictionary, or tuple whose elements can be serialized by Ray.
+
+For a more general object, Ray will first attempt to serialize the object by
+unpacking the object as a dictionary of its fields. This behavior is not
+correct in all cases. If Ray cannot serialize the object as a dictionary of its
+fields, Ray will fall back to using pickle. However, using pickle will likely
+be inefficient.
+
+
+Notes and limitations
```

Comment:
Given how many problems people have with serialization, it probably deserves its own documentation page. So we can show people how to use custom serializers and point out some of the current bugs (like serializing objects that contain duplicate copies of the same object).